ampd/src/starknet/events/contract_call.rs

```diff
@@ -0,0 +1,278 @@
+use std::num::TryFromIntError;
+
+use ethers::types::H256;
+use starknet_core::types::ValueOutOfRangeError;
+use starknet_core::utils::{parse_cairo_short_string, ParseCairoShortStringError};
+use thiserror::Error;
+
+use crate::starknet::events::EventType;
+use crate::starknet::types::byte_array::{ByteArray, ByteArrayError};
+use crate::types::Hash;
+
+/// This is the event emitted by the gateway cairo contract on Starknet,
+/// when the call_contract method is called from a third party.
+#[derive(Debug, PartialEq)]
+pub struct ContractCallEvent {
+    pub destination_address: String,
+    pub destination_chain: String,
+    pub source_address: String,
+    pub payload_hash: Hash,
+}
+
+/// An error, representing failure to convert/parse a starknet event
+/// to some specific event.
+#[derive(Error, Debug)]
+pub enum ContractCallError {
+    #[error("Invalid ContractCall event: {0}")]
+    InvalidEvent(String),
+    #[error("Cairo short string parse error: {0}")]
+    Cairo(#[from] ParseCairoShortStringError),
+    #[error("FieldElement operation errored with out of range: {0}")]
+    FeltOutOfRange(#[from] ValueOutOfRangeError),
+    #[error("Failed int conversion: {0}")]
+    TryFromConversion(#[from] TryFromIntError),
+    #[error("ByteArray type error: {0}")]
+    ByteArray(#[from] ByteArrayError),
+}
+
+impl TryFrom<starknet_core::types::Event> for ContractCallEvent {
+    type Error = ContractCallError;
+
+    fn try_from(starknet_event: starknet_core::types::Event) -> Result<Self, Self::Error> {
```

Comment:
suggestion: this method could be broken down into smaller bits I think

Reply:
The only meaningful breakdown I can think of, is to create separate methods for every part of the contract_call.
But I don't think they'll be very useful by their own or be able to be reused.

I'll think a bit more, but we can also skip it, I think.